design-docs/001-virtual-chunk-improvements.md

```diff
@@ -0,0 +1,163 @@
+# Improvements to virtual chunks handling
+
+## What we need
+
+1. Virtual chunks can be stored in all supported object stores
+2. Multiple credentials sets for virtual chunk resolution
```

Comment:
I think it would be useful to create virtual stores that would allow people to open a data tree containing both raw data and overviews. The raw data would commonly be in a separate bucket from the overviews.

Reply:
> Can you elaborate? Why? Could the overview be native writes to the repo?

To be fair, I'm not sure how common this is but the primary case in which the overviews cannot be native writes to the repo is when they already exist as Zarr V2 data.

A more common scenario would be continuously adding NetCDF references as virtual chunks along with continuously adding Zarr overviews. My understanding from previous conversations was that once native data is committed to a repository, all new commits will need to be native rather than virtual. Is that incorrect? Could one continuously add virtual chunks for the full resolution data and native chunks for the overviews?

Reply:
>  Could one continuously add virtual chunks for the full resolution data and native chunks for the overviews?

Icechunk does not place restrictions on which chunks are native and which are virtual

Reply:
> My understanding from previous conversations was that once native data is committed to a repository, all new commits will need to be native rather than virtual. Is that incorrect

That's incorrect. You can mix native and virtual chunks however you want.

In the scenario you describe, I'd want to make the overviews native writes and only use virtual chunks for the original full-resolution data.

I understand that my proposal (only one location for virtual chunks per repo) may be constraining, but that is deliberate.

Reply:
Thanks for the clarification about mixing native and virtual chunks.

> I understand that my proposal (only one location for virtual chunks per repo) may be constraining, but that is deliberate.

👍  Given the early design phase, it seemed worthwhile to detail what types of use-cases would be excluded in this design.

> 2. Virtual chunks within a repo can be stored in _one single object storage bucket in a specific cloud platform_. Mixing virtual references from multiple buckets in a single repo is not allowed.

For most of the data searchable via earthaccess, there are two storage buckets listed - a public and a private bucket (e.g., `s3://lp-prod-protected/MYD14.061`, `s3://lp-prod-public/MYD14.061`. Should this scenario require two separate virtual stores?

Reply:
> Icechunk does not place restrictions on which chunks are native and which are virtual

Yes you should already be able to create "mixed" icechunk stores using xarray and virtualizarr. [Aimee's `append_dim` PR](https://github.com/zarr-developers/VirtualiZarr/pull/272) goes even further because it is one way to create a mixed _array_, where you could append native chunks to an array containing virtual chunks (see https://github.com/zarr-developers/VirtualiZarr/pull/272#issuecomment-2483603659).

---

> @TomNicholas - can you think of scenarios where you would want to mix multiple different buckets in a single repo?

If you wanted to use virtual references to create some kind of aggregated dataset, e.g. put a bunch of disparate model run datasets that together comprise CMIP6 into one virtual store.

Those model runs could be in different buckets (or even clouds...) from different data providers. This kind of scenario is a little more likely to occur across groups in a store rather than within a single group/array IMO - think a `DataTree` where each group came from a different source. I think its debatable whether that use case should be targeted by some kind of higher-level catalog instead though.

Reply:
yes there is an extremely thin line between "catalog" and "manifest" with arbitrary trees.

Reply:
I think things look much better in the latest commit. Please take a look and tell me what you think. The solution seems much simpler, and still allows all kind of mixes of object stores, prefixes, credentials, etc.